File: src/main/Java/boj_step14/boj7785.java
import java.io.*;
import java.util.*;

public class boj7785 {
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
        int n = Integer.parseInt(reader.readLine());
        HashSet<String> list = new HashSet<>();
        while(n-- > 0) {
            StringTokenizer tokenizer = new StringTokenizer(reader.readLine());
            String name = tokenizer.nextToken();
            String clock = tokenizer.nextToken();
            if (clock.equals("enter")) list.add(name);
            else if (clock.equals("leave")) list.remove(name);
        }
        List<String> sortedList = new ArrayList<>(list);
        Collections.sort(sortedList, Comparator.reverseOrder());
        for (String name: sortedList) writer.append(name + "\n");
        writer.close();
    }
}

File: src/main/kotlin/boj_step14/boj7785.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter

fun main() {
    val r = BufferedReader(InputStreamReader(System.`in`))
    val w = BufferedWriter(OutputStreamWriter(System.out))
    val n = r.readLine().toInt()
    var map = mutableMapOf<String, String>()
    var list = mutableListOf<String>()
    repeat(n) {
        val (name, log) = r.readLine().split(" ").map { it }
        if (log == "enter") map.put(name, log)
        else if (log == "leave") map.remove(name)
    }
    map.forEach { name, log ->
        list.add(name)
    }
    list = list.sorted().reversed().toMutableList()
    list.forEach {
        w.write(it)
        w.newLine()
    }
    w.close()
}